### PR TITLE
Feature/mblinter 96 event owner functionallity

### PR DESCRIPTION
**Pull request by marija-drobac**
event owner functionality

### CODE REVIEW

File: app/src/main/java/com/example/hangoutz/ui/components/InputFieldWithIcon.kt
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/components/InputFieldWithIcon.kt:1]: Format file.

File: app/src/main/java/com/example/hangoutz/ui/components/InputField.kt
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/components/InputField.kt:1]: Format file.

File: app/src/main/java/com/example/hangoutz/ui/screens/events/EventScreenViewModel.kt [hunk 1]
```diff
@@ -225,4 +225,7 @@ class EventScreenViewModel @Inject constructor(
             }
         }
     }
+
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/events/EventScreenViewModel.kt]: Remove empty lines.

File: app/src/main/java/com/example/hangoutz/ui/screens/events/EventsScreen.kt [hunk 2]
```diff
@@ -214,11 +209,19 @@ fun EventsList(
...
-                3000,
-                easing = EaseInOut
+            360f, animationSpec = tween(
+                3000, easing = EaseInOut
             )
         )
     }
+}
+
+fun isCurrentUserOwner(context: Context, event: EventCardDPO): Boolean {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/events/EventsScreen.kt]: Move to view model or see on which page you are and then based on that navigate to different screens.

**Review by Mentors2024 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/hangoutz/ui/components/TimePicker.kt [hunk 3]
```diff
@@ -36,10 +36,9 @@ fun TimePickerModal(
         initialMinute = currentTime.get(Calendar.MINUTE),
         is24Hour = true,
     )
-
     BasicAlertDialog(
         onDismissRequest = onDismiss,
-        modifier = Modifier.clip(RoundedCornerShape(Dimensions.TIME_PICKER_CORNERS))
+        modifier = Modifier.clip(RoundedCornerShape(20.dp))
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/components/TimePicker.kt]: Use dp from dimensions file.

File: app/src/main/java/com/example/hangoutz/ui/screens/createEvent/CreateEventScreen.kt [hunk 4]
```diff
@@ -217,64 +214,60 @@ fun CreateEventScreen(
...
-            ActionButton(stringResource(R.string.event_create),
-                modifier = Modifier
-                    .padding(bottom = Dimensions.ACTION_BUTTON_MEDIUM3)
-                    .semantics {
-                        contentDescription = Constants.CREATE_EVENT_CREATE_BUTTON
-                    },
-                onClick = {
-                    viewmodel.createEvent()
-                })
-            if (sheetState.isVisible) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/createEvent/CreateEventScreen.kt]: Talk to Sasa this changes should stay.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 5]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+
         } else {
-            _uiState.value = _uiState.value.copy(errorMessage = Constants.ERROR_EMPTY_FIELD)
             Log.e("Error", "Fields marked with * cant be empty")
         }
     }
 
-    fun deleteEvent() {
-        //TODO
+    private fun formatForDatabase() {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Move date formatting to Extensions file.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 6]
```diff
@@ -34,8 +33,21 @@ data class EventDetailsData(
     var participants: List<User> = emptyList(),
     var showDatePicker: Boolean = false,
     var showTimePicker: Boolean = false,
-    var isError: Boolean = false,
-    var errorMessage: String? = null
+    var isTitleError: Boolean = false,
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt:47]: Do we need all this errors even for the fields that can be empty?

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 7]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+        getData()
     }
 
     private fun formatDateTime(dateTimeString: String): Pair<String, String> {
 
-        val inputFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault())
-        val dateFormat = SimpleDateFormat("dd.MM.yyyy", Locale.getDefault())
+        val inputFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm", Locale.getDefault())
+        val dateFormat = SimpleDateFormat("dd.MM.yyyy.", Locale.getDefault())
         val timeFormat = SimpleDateFormat("HH.mm", Locale.getDefault())
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Time format should be HH:mm not HH.mm.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 8]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+            if (response?.isSuccessful == true) {
+                Log.i("RemoveUser", "User successfully removed from event")
+            } else {
+                Log.e("RemoveUser", "Failed to remove user from event")
+            }
+        }
+        getData()
     }
 
     private fun formatDateTime(dateTimeString: String): Pair<String, String> {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Move this function to Extensions.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 9]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
 
-    fun validateInputs(): Boolean {
-        //TODO - field validation
-        return false
+    fun checkLength(text: String, length: Int): Boolean {
+        if (text.length <= length) return true
+        else return false
+    }
+
+    fun checkIfInPast(date: String): Boolean {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Rename to isEventPassed.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 10]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+        var isValid: Boolean = false
+        val inputFormat = SimpleDateFormat("dd.MM.yyyy. HH.mm", Locale.getDefault())
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Remove . at the end of year.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 11]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+        var validateTime = _uiState.value.time.trim().isEmpty()
+        var validateDesc = false
+        var validateCity = false
+        var validateStreet = false
+
+        if (validateTitle || validatePlace || validateDate || validateTime) {
+            _uiState.value = _uiState.value.copy(errorMessage = Constants.ERROR_EMPTY_FIELD)
+        }
+
+        if (!validateTitle && !checkLength(_uiState.value.title, 25)) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Move to Constants or use from constants if already exists.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 12]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+            validateTitle = true
+            _uiState.value = _uiState.value.copy(errorTitle = Constants.ERROR_TOO_LONG)
+        }
+
+        if (!validatePlace && !checkLength(_uiState.value.place, 25)) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Same here.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 13]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+            val combinedInput = "${_uiState.value.date} ${_uiState.value.time}"
+            if (checkIfInPast(combinedInput)) {
+                validateDate = true
+                validateTime = true
+                _uiState.value =
+                    _uiState.value.copy(errorMessage = "Date and time cannot be in the past")
+            }
+        }
+
+        if (_uiState.value.description.length > 100) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: And here.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 14]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+            validateDesc = true
+            _uiState.value = _uiState.value.copy(errorDesc = Constants.ERROR_TOO_LONG_DESC)
+        }
+
+        if (_uiState.value.city.length > 25) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: And here.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 15]
```diff
@@ -125,9 +204,94 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+            validateCity = true
+            _uiState.value = _uiState.value.copy(errorCity = Constants.ERROR_TOO_LONG)
+        }
+
+        if (_uiState.value.street.length > 25) {
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: And also here.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 16]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
     }
 
     fun editEvent(onSuccess: () -> Unit) {
+        formatForDatabase()
+        Log.e(
+            "DATA",
+            "id=  ${_uiState.value.eventId.toString()}, newTitle =  ${_uiState.value.title}, newPlace = ${_uiState.value.place}, newDate =  ${_uiState.value.formattedDateForDatabase}"
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Add space after id.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 17]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
-        //TODO
+    private fun formatForDatabase() {
+        val inputDate = _uiState.value.date
+        val inputTime = _uiState.value.time
+
+        val combined = "$inputDate $inputTime"
+
+        val inputFormat = SimpleDateFormat("dd.MM.yyyy. HH.mm", Locale.getDefault())
+        val outputFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
+        val formattedDateTime = outputFormat.format(inputFormat.parse(combined)!!)
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt]: Remove !!.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt [hunk 18]
```diff
@@ -57,26 +69,92 @@ class EventDetailsOwnerViewModel @Inject constructor(
...
+        Log.e(
+            "DATA",
+            "id=  ${_uiState.value.eventId.toString()}, newTitle =  ${_uiState.value.title}, newPlace = ${_uiState.value.place}, newDate =  ${_uiState.value.formattedDateForDatabase}"
+        )
         if (validateInputs()) {
-            onSuccess()
+            _uiState.value = _uiState.value.copy(errorMessage = "")
+            viewModelScope.launch {
+
+                val eventResponse = eventRepository.patchEventById(
```
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt:96]: We should pass all fields not just required fields. This way if you update for example description event wont be updated.

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt
Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt:1]: When we open calendar and time popup should have values from the event not default ones.

Comment (Mentors2024) [inline app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt:1]: We should be able to delete event even if we have participants. When we delete event first we need to delete all participants and then we can delete event.

**Review by Mentors2024 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt (hunk 6 above)
Comment (marija-drobac) [reply app/src/main/java/com/example/hangoutz/ui/screens/eventDetailsOwner/EventDetailsOwnerViewModel.kt:47]: yes, because the length cant me more than 25

**Review by Mentors2024 — APPROVED** [review]

**Pull request merged**